File: templates/Uwp/_composition/_shared/Feature.ShareTarget/Helpers/ShareOperationExtensions.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Windows.ApplicationModel.DataTransfer;
using Windows.ApplicationModel.DataTransfer.ShareTarget;
using Windows.Storage.Streams;
using Windows.Storage;

namespace Param_RootNamespace.Helpers
{
    public static class ShareOperationExtensions
    {
        public static async Task<Uri> GetApplicationLinkAsync(this ShareOperation shareOperation)
        {
            return await GetOperationDataAsync<Uri>(shareOperation, StandardDataFormats.ApplicationLink);
        }

        public static async Task<RandomAccessStreamReference> GetBitmapAsync(this ShareOperation shareOperation)
        {
            return await GetOperationDataAsync<RandomAccessStreamReference>(shareOperation, StandardDataFormats.Bitmap);
        }

        public static async Task<string> GetHtmlFormatAsync(this ShareOperation shareOperation)
        {
            return await GetOperationDataAsync<string>(shareOperation, StandardDataFormats.Html);
        }

        public static async Task<string> GetRtfAsync(this ShareOperation shareOperation)
        {
            return await GetOperationDataAsync<string>(shareOperation, StandardDataFormats.Rtf);
        }

        public static async Task<IReadOnlyList<IStorageItem>> GetStorageItemsAsync(this ShareOperation shareOperation)
        {
            return await GetOperationDataAsync<IReadOnlyList<IStorageItem>>(shareOperation, StandardDataFormats.StorageItems);
        }

        public static async Task<string> GetTextAsync(this ShareOperation shareOperation)
        {
            return await GetOperationDataAsync<string>(shareOperation, StandardDataFormats.Text);
        }

        public static async Task<Uri> GetWebLinkAsync(this ShareOperation shareOperation)
        {
            return await GetOperationDataAsync<Uri>(shareOperation, StandardDataFormats.WebLink) as Uri;
        }

        private static async Task<T> GetOperationDataAsync<T>(this ShareOperation shareOperation, string dataFormat)
            where T : class
        {
            try
            {
                if (dataFormat == StandardDataFormats.ApplicationLink)
                {
                    return await shareOperation.Data.GetApplicationLinkAsync() as T;
                }

                if (dataFormat == StandardDataFormats.Bitmap)
                {
                    return await shareOperation.Data.GetBitmapAsync() as T;
                }

                if (dataFormat == StandardDataFormats.Html)
                {
                    return await shareOperation.Data.GetHtmlFormatAsync() as T;
                }

                if (dataFormat == StandardDataFormats.Rtf)
                {
                    return await shareOperation.Data.GetRtfAsync() as T;
                }

                if (dataFormat == StandardDataFormats.StorageItems)
                {
                    return await shareOperation.Data.GetStorageItemsAsync() as T;
                }

                if (dataFormat == StandardDataFormats.Text)
                {
                    return await shareOperation.Data.GetTextAsync() as T;
                }

                if (dataFormat == StandardDataFormats.WebLink)
                {
                    return await shareOperation.Data.GetWebLinkAsync() as T;
                }
            }
            catch (Exception)
            {
                return default(T);
            }

            return default(T);
        }
    }
}

File: templates/Uwp/_comp/_shared/Feat.ShareTarget._VB/Helpers/ShareOperationExtensions.vb
Imports Windows.ApplicationModel.DataTransfer
Imports Windows.ApplicationModel.DataTransfer.ShareTarget
Imports Windows.Storage.Streams
Imports Windows.Storage

Namespace Helpers
    Public Module ShareOperationExtensions

        <Extension>
        Public Async Function GetApplicationLinkAsync(shareOperation As ShareOperation) As Task(Of Uri)
            Return Await GetOperationDataAsync(Of Uri)(shareOperation, StandardDataFormats.ApplicationLink)
        End Function

        <Extension>
        Public Async Function GetBitmapAsync(shareOperation As ShareOperation) As Task(Of RandomAccessStreamReference)
            Return Await GetOperationDataAsync(Of RandomAccessStreamReference)(shareOperation, StandardDataFormats.Bitmap)
        End Function

        <Extension>
        Public Async Function GetHtmlFormatAsync(shareOperation As ShareOperation) As Task(Of String)
            Return Await GetOperationDataAsync(Of String)(shareOperation, StandardDataFormats.Html)
        End Function

        <Extension>
        Public Async Function GetRtfAsync(shareOperation As ShareOperation) As Task(Of String)
            Return Await GetOperationDataAsync(Of String)(shareOperation, StandardDataFormats.Rtf)
        End Function

        <Extension>
        Public Async Function GetStorageItemsAsync(shareOperation As ShareOperation) As Task(Of IReadOnlyList(Of IStorageItem))
            Return Await GetOperationDataAsync(Of IReadOnlyList(Of IStorageItem))(shareOperation, StandardDataFormats.StorageItems)
        End Function

        <Extension>
        Public Async Function GetTextAsync(shareOperation As ShareOperation) As Task(Of String)
            Return Await GetOperationDataAsync(Of String)(shareOperation, StandardDataFormats.Text)
        End Function

        <Extension>
        Public Async Function GetWebLinkAsync(shareOperation As ShareOperation) As Task(Of Uri)
            Return TryCast(Await GetOperationDataAsync(Of Uri)(shareOperation, StandardDataFormats.WebLink), Uri)
        End Function

        Private Async Function GetOperationDataAsync(Of T As Class)(shareOperation As ShareOperation, dataFormat As String) As Task(Of T)
            Try
                If dataFormat = StandardDataFormats.ApplicationLink Then
                    Return TryCast(Await shareOperation.Data.GetApplicationLinkAsync(), T)
                End If

                If dataFormat = StandardDataFormats.Bitmap Then
                    Return TryCast(Await shareOperation.Data.GetBitmapAsync(), T)
                End If

                If dataFormat = StandardDataFormats.Html Then
                    Return TryCast(Await shareOperation.Data.GetHtmlFormatAsync(), T)
                End If

                If dataFormat = StandardDataFormats.Rtf Then
                    Return TryCast(Await shareOperation.Data.GetRtfAsync(), T)
                End If

                If dataFormat = StandardDataFormats.StorageItems Then
                    Return TryCast(Await shareOperation.Data.GetStorageItemsAsync(), T)
                End If

                If dataFormat = StandardDataFormats.Text Then
                    Return TryCast(Await shareOperation.Data.GetTextAsync(), T)
                End If

                If dataFormat = StandardDataFormats.WebLink Then
                    Return TryCast(Await shareOperation.Data.GetWebLinkAsync(), T)
                End If
            Catch generatedExceptionName As Exception
                Return Nothing
            End Try

            Return Nothing
        End Function
    End Module
End Namespace
